marshal.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+)
+
+// Marshal takes the input structure and creates an http request.
+//
+// See: Unmarshal for more details.
+//
+// For fields with a "path" item in the structural tag, the base uri must
+// contain a placeholder with its name.
+// Example:
+// For
+//    type Test struct {
+//	    username string `httprequest:"user,path"`
+//    }
+// ...the request url must contain a ":user" placeholder:
+//    http://localhost:8081/:user/files
+//
+// If a type does not implement the encoding.TextMarshaler fmt.Sprint will
+// be used to marshal its value.
+func Marshal(baseURL, method string, x interface{}) (*http.Request, error) {
+	xv := reflect.ValueOf(x)
+	pt, err := getRequestType(xv.Type())
+	if err != nil {
+		return nil, errgo.WithCausef(err, ErrBadUnmarshalType, "bad type %s", xv.Type())
+	}
+	req, err := http.NewRequest(method, baseURL, bytes.NewBuffer(nil))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	p := &Params{req, httprouter.Params{}}
+	if err := marshal(p, xv, pt); err != nil {
+		return nil, errgo.Mask(err, errgo.Is(ErrUnmarshal))
+	}
+	return p.Request, nil
+}
+
+// marshal is the internal version of Marshal.
+func marshal(p *Params, xv reflect.Value, pt *requestType) error {
+	if xv.Kind() == reflect.Ptr {
+		xv = xv.Elem()
+	}
+	for _, f := range pt.fields {
+		fv := xv.FieldByIndex(f.index)
+
+		// TODO store the field name in the field so
+		// that we can produce a nice error message.
+		if err := f.marshal(fv, p); err != nil {
+			return errgo.WithCausef(err, ErrUnmarshal, "cannot marshal field")
+		}
+	}
+
+	urlString := p.URL.Path
+	var pathBuffer bytes.Buffer
+	paramsByName := make(map[string]string)
```

Comment:
I'm not convinced that it's worthwhile building this map. There are likely
to be very few variable components in the typical URL path
and building the map is itself relatively costly. I'd suggest just using
p.PathVar.ByName.

In fact, how about factoring out the substitution into separate
a separate function. Something like this, perhaps?

	// marshal is the internal version of Marshal.
	func marshal(p *Params, xv reflect.Value, pt *requestType) error {
		xv = xv.Elem()
	
		for _, f := range pt.fields {
			fv := xv.FieldByIndex(f.index)
	
			// TODO store the field name in the field so
			// that we can produce a nice error message.
			if err := f.marshal(fv, p); err != nil {
				return errgo.WithCausef(err, ErrUnmarshal, "cannot marshal field")
			}
		}
		path, err := buildPath(p.URL.Path, p.PathVar)
		if err != nil {
			return errgo.Mask(err)
		}
		p.URL.Path = path
		p.URL.RawQuery = p.Form.Encode()
		return nil
	}
	
	func buildPath(path string, p httprouter.Params) (string, error) {
		pathBytes := make([]byte, 0, len(path)*2)
		for {
			s, rest := nextPathSegment(path)
			if s == "" {
				break
			}
			if s[0] != ':' && s[0] != '*' {
				pathBytes = append(pathBytes, s...)
				path = rest
				continue
			}
			if s[0] == '*' && rest != "" {
				return "", errgo.New("star path parameter is not at end of path")
			}
			if len(s) == 1 {
				return "", errgo.New("empty path parameter")
			}
			val := p.ByName(s[1:])
			if val == "" {
				return "", errgo.Newf("missing value for path parameter %q", s[1:])
			}
			if s[0] == '*' {
				if !strings.HasPrefix(val, "/") {
					return "", errgo.Newf("value %q for path parameter %q does not start with required /", val, s)
				}
				val = val[1:]
			}
			pathBytes = append(pathBytes, val...)
			path = rest
		}
		return string(pathBytes), nil
	}

	// nextPathSegment returns the next wildcard or constant
        // segment of the given path and everything after that
        // segment.
	func nextPathSegment(s string) (string, string) {
		if s == "" {
			return "", ""
		}
		if s[0] == ':' || s[0] == '*' {
			if i := strings.Index(s, "/"); i != -1 {
				return s[0:i], s[i:]
			}
			return s, ""
		}
		if i := strings.IndexAny(s, ":*"); i != -1 {
			return s[0:i], s[i:]
		}
		return s, ""
	}